subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestClassExecuter.java

```diff
@@ -129,13 +138,23 @@ private boolean allTestsFiltered(Runner runner, List<Filter> filters) {
 
         private final TestSelectionMatcher matcher;
 
-        public MethodNameFilter(Iterable<String> includedTests) {
-            matcher = new TestSelectionMatcher(includedTests);
+        public MethodNameFilter(TestSelectionMatcher matcher) {
+            this.matcher = matcher;
         }
 
         @Override
         public boolean shouldRun(Description description) {
-            return matcher.matchesTest(JUnitTestEventAdapter.className(description), JUnitTestEventAdapter.methodName(description));
+            if (matcher.matchesTest(JUnitTestEventAdapter.className(description), JUnitTestEventAdapter.methodName(description))) {
+                return true;
+            }
+
+            for (Description child : description.getChildren()) {
```

Comment:
I don't see this logic tested anywhere at the moment. Make sure that a test is skipped if none of its children are matched.